### PR TITLE
fsck replicas

### PR DESCRIPTION
**Pull request by kmlebedev**
check all replicas
and select volume id

### CODE REVIEW

File: weed/shell/command_volume_fsck.go [hunk 1]
```diff
@@ -327,19 +349,21 @@ func (c *commandVolumeFsck) collectOneVolumeFileIds(tempFolder string, volumeId
...
-		dst, openErr := os.OpenFile(getFilerFileIdFile(tempFolder, vid), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
-		if openErr != nil {
-			return fmt.Errorf("failed to create file %s: %v", getFilerFileIdFile(tempFolder, vid), openErr)
+	for _, volumeIdToServer := range dataNodeVolumeIdToVInfo {
+		for vid := range volumeIdToServer {
+			dst, openErr := os.OpenFile(getFilerFileIdFile(tempFolder, vid), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+			if openErr != nil {
+				return fmt.Errorf("failed to create file %s: %v", getFilerFileIdFile(tempFolder, vid), openErr)
+			}
+			files[vid] = dst
```
Comment (chrislusf) [inline weed/shell/command_volume_fsck.go:402]: one replicated volume id will map to multiple dst files here.

Comment (kmlebedev) [reply weed/shell/command_volume_fsck.go:402]: @chrislusf  This one set of records for the volume id from the filer will be mapped to all replicas of the volume idx files

File: weed/shell/command_volume_fsck.go [hunk 2]
```diff
@@ -210,66 +231,68 @@ func (c *commandVolumeFsck) collectFilerFileIdAndPaths(volumeIdToServer map[uint
...
+
+				fmt.Fprintf(writer, "marked %d on server %v writable for forced purge\n", volumeId, vinfo.server)
+
+				if verbose {
+					fmt.Fprintf(writer, "purging files from volume %d\n", volumeId)
+				}
 
-			if err := c.purgeFileIdsForOneVolume(volumeId, orphanFileIds, writer); err != nil {
-				return fmt.Errorf("purging volume %d: %v", volumeId, err)
+				if err := c.purgeFileIdsForOneVolume(volumeId, orphanFileIds, writer); err != nil {
```
Comment (chrislusf) [inline weed/shell/command_volume_fsck.go:329]: This also will repeatedly processing one volumeId.

I wonder how to test the change properly.

Comment (kmlebedev) [reply weed/shell/command_volume_fsck.go:329]: > I wonder how to test the change properly.

done

**Comment by kmlebedev** [comment]
As a result, after a couple of attempts it turned out

```
volume.fsck -v -findMissingChunksInFilerPath /buckets/doc -findMissingChunksInVolumeId 123 -reallyDeleteFromVolume
working directory: /tmp/sw_fsck825113600
collecting volume id and locations from master ...
collected 4 volumes and locations.
collecting volume 123 file ids from fast-volume-3.8080 ...
collecting volume 123 file ids from fast-volume-2:8080 ...
collecting file ids from filer ...
dataNode:fast-volume-2:8080	volume:123	entries:4	orphan:3	75.00%	0B
123,83ea00000000
123,a13900000000
123,5e5d000000000
```

**Comment by kmlebedev** [comment]
@chrislusf I think I've checked everything.

File: weed/shell/command_volume_fsck.go (hunk 1 above)
Comment (chrislusf) [reply weed/shell/command_volume_fsck.go:402]: will this `dst` file be opened multiple times? but only closed once?

Comment (kmlebedev) [reply weed/shell/command_volume_fsck.go:402]: > will this `dst` file be opened multiple times? but only closed once?

no it is checked above
https://github.com/chrislusf/seaweedfs/pull/2858/files#diff-8e0b202ea9cb980864a59d8e563827cb5748abc6ed60be322ccca4b49abb7a94R168

**Pull request merged**